diff --git a/CMakeLists.txt b/CMakeLists.txt
index b77fb64a1..520b65eaa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -585,7 +585,8 @@ ELSEIF(OSG_GLES3_AVAILABLE)
         SET(OPENGL_HEADER1 "#include \"TargetConditionals.h\"" CACHE STRING "#include<> line for OpenGL Header")
         SET(OPENGL_HEADER2 "#include <OpenGLES/ES3/gl.h>" CACHE STRING "#include<> line for additional OpenGL Headers if required")
     ELSE()
-        SET(OPENGL_HEADER1 "#include <GLES3/gl3.h>" CACHE STRING "#include<> line for OpenGL Header")
+        #SET(OPENGL_HEADER1 "#include <GLES3/gl3.h>" CACHE STRING "#include<> line for OpenGL Header")
+        SET(OPENGL_HEADER1 "#include <glad/gles2.h>" CACHE STRING "#include<> line for OpenGL Header")
         SET(OPENGL_HEADER2 "" CACHE STRING "#include<> line for additional OpenGL Headers if required")
     ENDIF()
 ELSE()
@@ -652,10 +653,10 @@ IF(ANDROID)
                  PATHS
                      ${ANDROID_SYSROOT}/usr/lib)
     ELSEIF(OSG_GLES3_AVAILABLE)
-        FIND_PATH(OPENGL_INCLUDE_DIR GLES3/gl3.h
+        FIND_PATH(OPENGL_INCLUDE_DIR glad/gles2.h
               PATHS
               ${ANDROID_SYSROOT}/usr/include)
-        FIND_LIBRARY(OPENGL_gl_LIBRARY  GLESv3
+        FIND_LIBRARY(OPENGL_gl_LIBRARY  glad
                  PATHS
                      ${ANDROID_SYSROOT}/usr/lib)
     ENDIF()
