diff --git a/apps/openmw/mwrender/water.cpp b/apps/openmw/mwrender/water.cpp
index 584cd09d03..12e0000d92 100644
--- a/apps/openmw/mwrender/water.cpp
+++ b/apps/openmw/mwrender/water.cpp
@@ -85,7 +85,7 @@ namespace MWRender
                 if (true)
                 {
                     stateset = new osg::StateSet;
-                    SceneUtil::setClipPlane(*stateset, 0, osg::Plane(plane.getNormal(), 0).asVec4());
+                    SceneUtil::setClipPlane(*stateset, 0, plane.asVec4());
                 }
 
                 if (stateset)
diff --git a/components/shader/shadermanager.cpp b/components/shader/shadermanager.cpp
index 130c2dccd3..dbd79145dd 100644
--- a/components/shader/shadermanager.cpp
+++ b/components/shader/shadermanager.cpp
@@ -22,6 +22,7 @@
 #include <components/misc/strings/conversion.hpp>
 #include <components/sceneutil/clipplane.hpp>
 #include <components/settings/settings.hpp>
+#include <components/shader/shaderconverter.hpp>
 
 namespace
 {
@@ -45,6 +46,17 @@ namespace
         throw std::runtime_error("unrecognized shader template name: " + templateName);
     }
 
+    void applySpirvConversion(std::string& source, osg::Shader::Type type)
+    {
+        if (getenv("OPENMW_SPIRV_SHADERCONV") != nullptr)
+        {
+            //Log(Debug::Error) << "source before \n" << source << "\n\n";
+            Misc::StringUtils::replaceAll(source, "#version 320 es", "#version 460");
+            source = Shader::ShaderConverter::convertGLSLtoGLES(source, type, 320);
+            //Log(Debug::Error) << "source after \n" << source << "\n\n";
+        }
+    }
+
     std::string_view getRootPrefix(std::string_view path)
     {
         if (path.starts_with("lib"))
@@ -560,6 +572,8 @@ namespace Shader
                 return nullptr;
             }
 
+            applySpirvConversion(shaderSource, type ? *type : getShaderType(templateName));
+
             osg::ref_ptr<osg::Shader> shader(new osg::Shader(type ? *type : getShaderType(templateName)));
             shader->setShaderSource(shaderSource);
             // Assign a unique prefix to allow the SharedStateManager to compare shaders efficiently.
@@ -652,6 +666,9 @@ namespace Shader
                 // we would when creating the shader. If we put a nullptr in the shader map, we just lose the ability to
                 // put a working one in later.
                 continue;
+
+            applySpirvConversion(shaderSource, shader->getType());
+
             shader->setShaderSource(shaderSource);
 
             getLinkedShaders(shader, linkedShaderNames, defines);
diff --git a/files/shaders/core/ripples_blobber.comp b/files/shaders/core/ripples_blobber.comp
index 235f58a5ba..81e849b2f4 100644
--- a/files/shaders/core/ripples_blobber.comp
+++ b/files/shaders/core/ripples_blobber.comp
@@ -16,7 +16,7 @@ uniform vec2 offset;
 
 void main()
 {
-    ivec2 texel = ivec2(gl_GlobalInvocationID.xy + offset);
+    ivec2 texel = ivec2(gl_GlobalInvocationID.xy + uvec2(offset));
 
     vec4 color = imageLoad(imageIn, texel);
     float wavesizeMultiplier = getTemporalWaveSizeMultiplier(osg_SimulationTime);
diff --git a/files/shaders/lib/core/clip.glsl b/files/shaders/lib/core/clip.glsl
index 126771ef17..5b4827b7a0 100644
--- a/files/shaders/lib/core/clip.glsl
+++ b/files/shaders/lib/core/clip.glsl
@@ -14,7 +14,7 @@ void applyClipPlanes(vec4 pos)
 #ifdef GL_EXT_clip_cull_distance
         if (isReflection) gl_ClipDistance[@clipPlaneIndex] = dot(clipPlane@clipPlaneIndex, pos);
 #else
-        clip = dot(clipPlane@clipPlaneIndex, pos);
+        if (isReflection) clip = dot(clipPlane@clipPlaneIndex, pos);
 #endif
 
     @endforeach
