diff --git a/apps/openmw/engine.cpp b/apps/openmw/engine.cpp
index 6e4a2d2f8d..94c6291734 100644
--- a/apps/openmw/engine.cpp
+++ b/apps/openmw/engine.cpp
@@ -623,6 +623,9 @@ void OMW::Engine::setSkipMenu(bool skipMenu, bool newGame)
     mNewGame = newGame;
 }
 
+// To share the viewer with Android interfaces
+osg::ref_ptr<osgViewer::Viewer> g_viewer;
+
 void OMW::Engine::createWindow(bool hidden)
 {
     const int screen = Settings::video().mScreen;
@@ -875,6 +878,9 @@ void OMW::Engine::createWindow(bool hidden)
 
     mViewer->getEventQueue()->getCurrentEventState()->setWindowRectangle(
         0, 0, graphicsWindow->getTraits()->width, graphicsWindow->getTraits()->height);
+
+    // To share the viewer with Android interfaces
+    g_viewer = mViewer;
 }
 
 bool OMW::Engine::createVulkanWindow(std::uint32_t flags, int posX, int posY, Settings::RendererEntry entry)
@@ -1807,6 +1813,8 @@ void OMW::Engine::go()
 
     mLuaWorker->join();
 
+    g_viewer.release();
+
     // Save user settings
     Settings::Manager::saveUser(mCfgMgr.getUserConfigPath() / "settings.cfg");
     Settings::ShaderManager::get().save();
diff --git a/files/data/mygui/openmw_postprocessor_hud.layout b/files/data/mygui/openmw_postprocessor_hud.layout
index 4ed8ecdb36..758b1b2e99 100644
--- a/files/data/mygui/openmw_postprocessor_hud.layout
+++ b/files/data/mygui/openmw_postprocessor_hud.layout
@@ -3,7 +3,7 @@
   <Widget type="Window" skin="MW_Window" position="0 0 500 500" layer="Debug" name="_Main" align="Stretch">
     <Property key="Visible" value="false"/>
     <Property key="Caption" value="#{OMWShaders:PostProcessHUD}"/>
-    <Property key="MinSize" value="600 600"/>
+    <Property key="MinSize" value="600 430"/>
 
     <Widget type="TabControl" skin="TabControl_NoBorder" position="8 8 468 450" align="Stretch" name="TabControl">
         <Property key="ButtonAutoWidth" value="true"/>
@@ -20,7 +20,7 @@
             <Property key="Padding" value="0"/>
             <Property key="Spacing" value="8"/>
 
-            <Widget type="Widget" position ="0 0 0 300">
+            <Widget type="Widget" position ="0 0 0 250">
                 <UserString key="HStretch" value="true"/>
                 <UserString key="VStretch" value="false"/>
 
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e6c5814cef..c4d93e0dbd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -703,7 +703,7 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC" OR CMAKE_CXX_COMPILER_FRONTEND_VARIANT
         add_compile_options("/WX")
     endif()
 else ()
-    add_compile_options("-Wall" "-Wextra" "-Wundef" "-Wextra-semi" "-Wno-unused-parameter" "-pedantic" "-Wno-long-long" "-Wnon-virtual-dtor" "-Wunused" "-Wshadow")
+    add_compile_options("-Wall" "-Wextra" "-Wundef" "-Wextra-semi" "-Wno-unused-parameter" "-pedantic" "-Wno-long-long" "-Wnon-virtual-dtor" "-Wunused" "-Wshadow" "-Wno-missing-template-arg-list-after-template-kw")
 
     if (CMAKE_CXX_COMPILER_ID STREQUAL Clang AND NOT APPLE)
         if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 3.6 OR CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL 3.6)
diff --git a/apps/openmw/mwrender/clockwork/frame/postchain.cpp b/apps/openmw/mwrender/clockwork/frame/postchain.cpp
index 38102f64a1..1116fc5f65 100644
--- a/apps/openmw/mwrender/clockwork/frame/postchain.cpp
+++ b/apps/openmw/mwrender/clockwork/frame/postchain.cpp
@@ -381,7 +381,7 @@ namespace MWRender::Clockwork
         for (const osg::ref_ptr<osg::Texture>& texture : technique.getTextures())
         {
             const GLenum target = texture->getTextureTarget();
-            const VkImageType type = target == GL_TEXTURE_1D
+            const VkImageType type = target == 0x0DE0/*GL_TEXTURE_1D*/
                 ? VK_IMAGE_TYPE_1D
                 : (target == GL_TEXTURE_3D ? VK_IMAGE_TYPE_3D : VK_IMAGE_TYPE_2D);
             result.mTextures.push_back({ texture->getImage(0), samplerOptions(*texture), type });
