diff --git a/apps/opencs/view/render/scenewidget.cpp b/apps/opencs/view/render/scenewidget.cpp
index 071127d6160b719d13694bb85ceebc2fbe328cdc..28847c013e0c52a11002aec05b79e56b7de7bab2 100644
--- a/apps/opencs/view/render/scenewidget.cpp
+++ b/apps/opencs/view/render/scenewidget.cpp
@@ -49,6 +49,7 @@
 #include <components/sceneutil/glextensions.hpp>
 #include <components/sceneutil/lightmanager.hpp>
 #include <components/sceneutil/stateupdater.hpp>
+#include <components/sceneutil/texmat.hpp>
 #include <components/shader/removedalphafunc.hpp>
 
 #include "../widget/scenetoolmode.hpp"
@@ -172,6 +173,8 @@ namespace CSVRender
         mRootNode->getOrCreateStateSet()->setMode(
             GL_LIGHTING, osg::StateAttribute::OFF | osg::StateAttribute::PROTECTED | osg::StateAttribute::OVERRIDE);
 
+        SceneUtil::initTexMatForStateSet(*mRootNode->getOrCreateStateSet());
+
         mView->getCamera()->setViewport(new osg::Viewport(0, 0, width(), height()));
 
         mView->getCamera()->getOrCreateStateSet()->setMode(GL_NORMALIZE, osg::StateAttribute::ON);
diff --git a/apps/openmw/mwrender/renderingmanager.cpp b/apps/openmw/mwrender/renderingmanager.cpp
index 56782ac55fdedf2a8d23c36463dac55a176cff56..af2101f91db23f81ed6c6db9f25860603cda421a 100644
--- a/apps/openmw/mwrender/renderingmanager.cpp
+++ b/apps/openmw/mwrender/renderingmanager.cpp
@@ -37,6 +37,7 @@
 #include <components/sceneutil/rtt.hpp>
 #include <components/sceneutil/shadow.hpp>
 #include <components/sceneutil/stateupdater.hpp>
+#include <components/sceneutil/texmat.hpp>
 #include <components/sceneutil/visitor.hpp>
 #include <components/sceneutil/workqueue.hpp>
 #include <components/sceneutil/writescene.hpp>
@@ -407,6 +408,8 @@ namespace MWRender
             mRootNode->getOrCreateStateSet()->setAttributeAndModes(clipcontrol, osg::StateAttribute::ON);
         }
 
+        SceneUtil::initTexMatForStateSet(*mViewer->getSceneData()->getOrCreateStateSet());
+
         mRootNode->getOrCreateStateSet()->setMode(
             GL_LIGHTING, osg::StateAttribute::OFF | osg::StateAttribute::PROTECTED | osg::StateAttribute::OVERRIDE);
 
diff --git a/apps/openmw/mwrender/skyutil.cpp b/apps/openmw/mwrender/skyutil.cpp
index f331dd4fa674381c004f856d330a40608f0e4e73..9debe0395c21472e520cc06966556b4c595de66b 100644
--- a/apps/openmw/mwrender/skyutil.cpp
+++ b/apps/openmw/mwrender/skyutil.cpp
@@ -11,7 +11,6 @@
 #include <osg/Material>
 #include <osg/OcclusionQueryNode>
 #include <osg/PositionAttitudeTransform>
-#include <osg/TexMat>
 #include <osg/Transform>
 #include <osg/observer_ptr>
 
@@ -27,6 +26,7 @@
 #include <components/resource/scenemanager.hpp>
 
 #include <components/sceneutil/depth.hpp>
+#include <components/sceneutil/texmat.hpp>
 #include <components/sceneutil/texturetype.hpp>
 
 #include <components/fallback/fallback.hpp>
@@ -482,8 +482,7 @@ namespace MWRender
         stateset->setAttribute(
             createAlphaTrackingUnlitMaterial(), osg::StateAttribute::ON | osg::StateAttribute::OVERRIDE);
 
-        osg::ref_ptr<osg::TexMat> texmat = new osg::TexMat;
-        stateset->setTextureAttributeAndModes(0, texmat);
+        SceneUtil::setupTexMatForStateSet(*stateset, 0, osg::Matrixf{});
 
         stateset->setTextureAttribute(0, mTexture, osg::StateAttribute::ON | osg::StateAttribute::OVERRIDE);
 
@@ -498,8 +497,7 @@ namespace MWRender
         osg::Material* mat = static_cast<osg::Material*>(stateset->getAttribute(osg::StateAttribute::MATERIAL));
         mat->setEmission(osg::Material::FRONT_AND_BACK, mEmissionColor);
 
-        osg::TexMat* texMat = static_cast<osg::TexMat*>(stateset->getTextureAttribute(0, osg::StateAttribute::TEXMAT));
-        texMat->setMatrix(mTexMat);
+        SceneUtil::setupTexMatForStateSet(*stateset, 0, mTexMat);
 
         stateset->getUniform("opacity")->set(mOpacity);
     }
diff --git a/components/CMakeLists.txt b/components/CMakeLists.txt
index 7e153728aa70d975a4ab39a29154c89cc8e60fce..8c50727305c4814e232209b6c32daabfb73a9d7a 100644
--- a/components/CMakeLists.txt
+++ b/components/CMakeLists.txt
@@ -141,7 +141,7 @@ add_component_dir (sceneutil
     lightmanager lightutil positionattitudetransform workqueue pathgridutil waterutil writescene serialize optimizer
     detourdebugdraw navmesh agentpath animblendrules shadow mwshadowtechnique recastmesh shadowsbin osgacontroller rtt
     screencapture depth color riggeometryosgaextension extradata unrefqueue lightcommon clearcolor
-    cullsafeboundsvisitor keyframe nodecallback textkeymap glextensions
+    cullsafeboundsvisitor keyframe nodecallback textkeymap glextensions texmat
     )
 
 add_component_dir (nif
diff --git a/components/myguiplatform/myguirendermanager.cpp b/components/myguiplatform/myguirendermanager.cpp
index de54119189ce5226705daf3f4603cdd0612cd31c..4fcf5d7d366280a9fc687cd8e5e3caf4372fe810 100644
--- a/components/myguiplatform/myguirendermanager.cpp
+++ b/components/myguiplatform/myguirendermanager.cpp
@@ -3,7 +3,6 @@
 #include <MyGUI_Timer.h>
 
 #include <osg/Drawable>
-#include <osg/TexMat>
 #include <osg/Texture2D>
 
 #include <osgViewer/Viewer>
diff --git a/components/nifosg/controller.cpp b/components/nifosg/controller.cpp
index a62f77f571e65666d4a02f0b8c10c4f3fbf56588..f98af4e7db77880c5fb0f4ee87a1b1a14e5c7c28 100644
--- a/components/nifosg/controller.cpp
+++ b/components/nifosg/controller.cpp
@@ -2,7 +2,6 @@
 
 #include <osg/Material>
 #include <osg/MatrixTransform>
-#include <osg/TexMat>
 #include <osg/Texture2D>
 
 #include <osgAnimation/Bone>
@@ -12,6 +11,7 @@
 #include <components/nif/data.hpp>
 #include <components/sceneutil/clone.hpp>
 #include <components/sceneutil/morphgeometry.hpp>
+#include <components/sceneutil/texmat.hpp>
 
 #include "matrixtransform.hpp"
 
@@ -309,9 +309,8 @@ namespace NifOsg
 
     void UVController::setDefaults(osg::StateSet* stateset)
     {
-        osg::ref_ptr<osg::TexMat> texMat(new osg::TexMat);
         for (unsigned int unit : mTextureUnits)
-            stateset->setTextureAttributeAndModes(unit, texMat, osg::StateAttribute::ON);
+            SceneUtil::setupTexMatForStateSet(*stateset, unit, osg::Matrixf{});
     }
 
     void UVController::apply(osg::StateSet* stateset, osg::NodeVisitor* nv)
@@ -333,11 +332,7 @@ namespace NifOsg
 
             // setting once is enough because all other texture units share the same TexMat (see setDefaults).
             if (!mTextureUnits.empty())
-            {
-                osg::TexMat* texMat = static_cast<osg::TexMat*>(
-                    stateset->getTextureAttribute(*mTextureUnits.begin(), osg::StateAttribute::TEXMAT));
-                texMat->setMatrix(mat);
-            }
+                SceneUtil::setupTexMatForStateSet(*stateset, *mTextureUnits.begin(), mat);
         }
     }
 
diff --git a/components/nifosg/nifloader.cpp b/components/nifosg/nifloader.cpp
index b6e4b17642f104bab0835125ea9610887054b80e..c16935aecbce7b4bb09600ecf875ffb54967decb 100644
--- a/components/nifosg/nifloader.cpp
+++ b/components/nifosg/nifloader.cpp
@@ -9,7 +9,6 @@
 #include <osg/Matrixf>
 #include <osg/Sequence>
 #include <osg/Switch>
-#include <osg/TexMat>
 #include <osg/ValueObject>
 
 #include <yaml-cpp/yaml.h>
@@ -56,6 +55,7 @@
 #include <components/sceneutil/morphgeometry.hpp>
 #include <components/sceneutil/riggeometry.hpp>
 #include <components/sceneutil/skeleton.hpp>
+#include <components/sceneutil/texmat.hpp>
 #include <components/sceneutil/texturetype.hpp>
 
 #include "autotransform.hpp"
@@ -2647,7 +2647,6 @@ namespace NifOsg
                         attachExternalTexture("diffuseMap", VFS::Path::toNormalized(texprop->mSourceTexture),
                             texprop->wrapS(), texprop->wrapT(), uvSet, stateset, boundTextures);
                         {
-                            osg::ref_ptr<osg::TexMat> texMat(new osg::TexMat);
                             // This handles 20.2.0.7 UV settings like 4.0.0.2 UV settings (see NifOsg::UVController)
                             // TODO: verify
                             osg::Vec3f uvOrigin(0.5f, 0.5f, 0.f);
@@ -2659,8 +2658,7 @@ namespace NifOsg
                             mat.preMultTranslate(-uvOrigin);
                             mat.setTrans(mat.getTrans() + uvTrans);
 
-                            texMat->setMatrix(mat);
-                            stateset->setTextureAttributeAndModes(texUnit, texMat, osg::StateAttribute::ON);
+                            SceneUtil::setupTexMatForStateSet(*stateset, texUnit, mat);
                         }
                     }
                     bool useFalloff = texprop->useFalloff();
diff --git a/components/sceneutil/texmat.cpp b/components/sceneutil/texmat.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..bf0ca5f518d9834719440b0ca75214035fa1d32b
--- /dev/null
+++ b/components/sceneutil/texmat.cpp
@@ -0,0 +1,27 @@
+#include <format>
+
+#include <osg/Matrixf>
+#include <osg/StateSet>
+#include <osg/Uniform>
+
+namespace SceneUtil
+{
+
+    osg::ref_ptr<osg::Uniform> createTexMatUniform(unsigned int texUnit, const osg::Matrixf& mat)
+    {
+        return new osg::Uniform(std::format("texMat{}", texUnit).c_str(), mat);
+    }
+
+    void setupTexMatForStateSet(osg::StateSet& stateset, unsigned int texUnit, const osg::Matrixf& mat)
+    {
+        stateset.getOrCreateUniform(std::format("texMat{}", texUnit), osg::Uniform::FLOAT_MAT4)->set(mat);
+    }
+
+    void initTexMatForStateSet(osg::StateSet& stateset)
+    {
+        constexpr unsigned int maxTexMat = 32;
+        for (unsigned int i = 0; i < maxTexMat; ++i)
+            setupTexMatForStateSet(stateset, i, osg::Matrixf{});
+    }
+
+}
diff --git a/components/sceneutil/texmat.hpp b/components/sceneutil/texmat.hpp
new file mode 100644
index 0000000000000000000000000000000000000000..33807b075373b2eb3299a1256e6c45b2dc14e561
--- /dev/null
+++ b/components/sceneutil/texmat.hpp
@@ -0,0 +1,24 @@
+#ifndef OPENMW_COMPONENTS_SCENEUTIL_TEXMAT_H
+#define OPENMW_COMPONENTS_SCENEUTIL_TEXMAT_H
+
+#include <osg/ref_ptr>
+
+namespace osg
+{
+    class Matrixf;
+    class Uniform;
+    class StateSet;
+}
+
+namespace SceneUtil
+{
+
+    osg::ref_ptr<osg::Uniform> createTexMatUniform(unsigned int texUnit, const osg::Matrixf& mat);
+
+    void setupTexMatForStateSet(osg::StateSet& stateset, unsigned int texUnit, const osg::Matrixf& mat);
+
+    void initTexMatForStateSet(osg::StateSet& stateset);
+
+}
+
+#endif
diff --git a/components/sdlutil/sdlcursormanager.cpp b/components/sdlutil/sdlcursormanager.cpp
index 9e6719da39bf856d5889183d7a34b3556ee954ef..6b4507621e1458c7bc0b4d284b47c1c02353737b 100644
--- a/components/sdlutil/sdlcursormanager.cpp
+++ b/components/sdlutil/sdlcursormanager.cpp
@@ -9,7 +9,6 @@
 
 #include <osg/Geometry>
 #include <osg/GraphicsContext>
-#include <osg/TexMat>
 #include <osg/Texture2D>
 #include <osg/Version>
 #include <osgViewer/GraphicsWindow>
diff --git a/components/terrain/material.cpp b/components/terrain/material.cpp
index 04d5ead810540783a4b925cecafc223db5151e50..d04be28d191b848ccb3fbb78978dcd84db210395 100644
--- a/components/terrain/material.cpp
+++ b/components/terrain/material.cpp
@@ -4,11 +4,11 @@
 #include <osg/Capability>
 #include <osg/Depth>
 #include <osg/Fog>
-#include <osg/TexMat>
 #include <osg/Texture2D>
 
 #include <components/resource/scenemanager.hpp>
 #include <components/sceneutil/depth.hpp>
+#include <components/sceneutil/texmat.hpp>
 #include <components/sceneutil/util.hpp>
 #include <components/shader/shadermanager.hpp>
 #include <components/stereo/stereomanager.hpp>
@@ -20,13 +20,13 @@ namespace
     class BlendmapTexMat
     {
     public:
-        static const osg::ref_ptr<osg::TexMat>& value(const int blendmapScale)
+        static const osg::ref_ptr<osg::Uniform>& value(const int blendmapScale)
         {
             static BlendmapTexMat instance;
             return instance.get(static_cast<float>(blendmapScale));
         }
 
-        const osg::ref_ptr<osg::TexMat>& get(const float blendmapScale)
+        const osg::ref_ptr<osg::Uniform>& get(const float blendmapScale)
         {
             const std::lock_guard<std::mutex> lock(mMutex);
             auto texMat = mTexMatMap.find(blendmapScale);
@@ -42,26 +42,26 @@ namespace
                 // blendmap, apparently.
                 matrix.preMultTranslate(osg::Vec3f(1.0f / blendmapScale / 4.0f, -1.0f / blendmapScale / 4.0f, 0.f));
 
-                texMat = mTexMatMap.emplace(blendmapScale, new osg::TexMat(matrix)).first;
+                texMat = mTexMatMap.emplace(blendmapScale, SceneUtil::createTexMatUniform(1, matrix)).first;
             }
             return texMat->second;
         }
 
     private:
         std::mutex mMutex;
-        std::map<float, osg::ref_ptr<osg::TexMat>> mTexMatMap;
+        std::map<float, osg::ref_ptr<osg::Uniform>> mTexMatMap;
     };
 
     class LayerTexMat
     {
     public:
-        static const osg::ref_ptr<osg::TexMat>& value(const float layerTileSize)
+        static const osg::ref_ptr<osg::Uniform>& value(const float layerTileSize)
         {
             static LayerTexMat instance;
             return instance.get(layerTileSize);
         }
 
-        const osg::ref_ptr<osg::TexMat>& get(const float layerTileSize)
+        const osg::ref_ptr<osg::Uniform>& get(const float layerTileSize)
         {
             const std::lock_guard<std::mutex> lock(mMutex);
             auto texMat = mTexMatMap.find(layerTileSize);
@@ -69,7 +69,8 @@ namespace
             {
                 texMat = mTexMatMap
                              .insert(std::make_pair(layerTileSize,
-                                 new osg::TexMat(osg::Matrix::scale(osg::Vec3f(layerTileSize, layerTileSize, 1.f)))))
+                                 SceneUtil::createTexMatUniform(
+                                     0, osg::Matrix::scale(osg::Vec3f(layerTileSize, layerTileSize, 1.f)))))
                              .first;
             }
             return texMat->second;
@@ -77,7 +78,7 @@ namespace
 
     private:
         std::mutex mMutex;
-        std::map<float, osg::ref_ptr<osg::TexMat>> mTexMatMap;
+        std::map<float, osg::ref_ptr<osg::Uniform>> mTexMatMap;
     };
 
     class EqualDepth
@@ -215,7 +216,7 @@ namespace Terrain
             stateset->addUniform(UniformCollection::value().mDiffuseMap);
 
             if (layerTileSize != 1.f)
-                stateset->setTextureAttributeAndModes(0, LayerTexMat::value(layerTileSize), osg::StateAttribute::ON);
+                stateset->addUniform(LayerTexMat::value(layerTileSize));
 
             if (!blendmaps.empty())
             {
@@ -223,7 +224,7 @@ namespace Terrain
 
                 stateset->setTextureAttributeAndModes(1, blendmap.get());
                 if (!esm4terrain)
-                    stateset->setTextureAttributeAndModes(1, BlendmapTexMat::value(blendmapScale));
+                    stateset->addUniform(BlendmapTexMat::value(blendmapScale));
                 stateset->addUniform(UniformCollection::value().mBlendMap);
             }
             if (isComposite)
diff --git a/files/shaders/compatibility/bs/default.vert b/files/shaders/compatibility/bs/default.vert
index 1055f195db52144054e627f612665240daac2737..b21fcda82d697da2b0245981ca9b32df138bb788 100644
--- a/files/shaders/compatibility/bs/default.vert
+++ b/files/shaders/compatibility/bs/default.vert
@@ -27,6 +27,10 @@ varying float linearDepth;
 varying vec3 passViewPos;
 varying vec3 passNormal;
 
+uniform mat4 texMat@diffuseMapUV;
+uniform mat4 texMat@emissiveMapUV;
+uniform mat4 texMat@normalMapUV;
+
 #include "lib/view/depth.glsl"
 
 #include "compatibility/vertexcolors.glsl"
@@ -51,15 +55,15 @@ void main(void)
 #endif
 
 #if @diffuseMap
-    diffuseMapUV = (gl_TextureMatrix[@diffuseMapUV] * gl_MultiTexCoord@diffuseMapUV).xy;
+    diffuseMapUV = (texMat@diffuseMapUV * vec4(gl_MultiTexCoord@diffuseMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @emissiveMap
-    emissiveMapUV = (gl_TextureMatrix[@emissiveMapUV] * gl_MultiTexCoord@emissiveMapUV).xy;
+    emissiveMapUV = (texMat@emissiveMapUV * vec4(gl_MultiTexCoord@emissiveMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @normalMap
-    normalMapUV = (gl_TextureMatrix[@normalMapUV] * gl_MultiTexCoord@normalMapUV).xy;
+    normalMapUV = (texMat@normalMapUV * vec4(gl_MultiTexCoord@normalMapUV.xyz, 1.0)).xy;
 #endif
 
 
diff --git a/files/shaders/compatibility/bs/nolighting.vert b/files/shaders/compatibility/bs/nolighting.vert
index bc0a4f20c62aebe0abd1dff5cf4272ca704244f6..d08c8e2bf2512e8f9fb09553570f4041aa1441ae 100644
--- a/files/shaders/compatibility/bs/nolighting.vert
+++ b/files/shaders/compatibility/bs/nolighting.vert
@@ -19,6 +19,8 @@ varying float passFalloff;
 uniform bool useFalloff;
 uniform vec4 falloffParams;
 
+uniform mat4 texMat@diffuseMapUV;
+
 #include "lib/view/depth.glsl"
 
 #include "compatibility/vertexcolors.glsl"
@@ -34,7 +36,7 @@ void main(void)
     linearDepth = getLinearDepth(gl_Position.z, viewPos.z);
 
 #if @diffuseMap
-    diffuseMapUV = (gl_TextureMatrix[@diffuseMapUV] * gl_MultiTexCoord@diffuseMapUV).xy;
+    diffuseMapUV = (texMat@diffuseMapUV * vec4(gl_MultiTexCoord@diffuseMapUV.xyz, 1.0)).xy;
 #endif
 
     passColor = gl_Color;
diff --git a/files/shaders/compatibility/depthclipped.vert b/files/shaders/compatibility/depthclipped.vert
index d348ea1a8a19bb17acae689ee2aa6f498c8abadc..b4ce737391354bed21d6995c336db0d0e4192b88 100644
--- a/files/shaders/compatibility/depthclipped.vert
+++ b/files/shaders/compatibility/depthclipped.vert
@@ -3,6 +3,8 @@
 varying vec2 diffuseMapUV;
 varying float alphaPassthrough;
 
+uniform mat4 texMat0;
+
 #include "lib/core/vertex.h.glsl"
 #include "vertexcolors.glsl"
 
@@ -18,5 +20,5 @@ void main()
     else
         alphaPassthrough = gl_FrontMaterial.diffuse.a;
 
-    diffuseMapUV = (gl_TextureMatrix[0] * gl_MultiTexCoord0).xy;
+    diffuseMapUV = (texMat0 * vec4(gl_MultiTexCoord0.xyz, 1.0)).xy;
 }
diff --git a/files/shaders/compatibility/groundcover.vert b/files/shaders/compatibility/groundcover.vert
index fce64521f797007aac568d64d2a8d4c06f02bd1b..336e2b2fb312edbfbb66196e874a6b087b1b1281 100644
--- a/files/shaders/compatibility/groundcover.vert
+++ b/files/shaders/compatibility/groundcover.vert
@@ -13,10 +13,12 @@ attribute vec3 aRotation;
 
 #if @diffuseMap
 varying vec2 diffuseMapUV;
+uniform mat4 texMat@diffuseMapUV;
 #endif
 
 #if @normalMap
 varying vec2 normalMapUV;
+uniform mat4 texMat@normalMapUV;
 #endif
 
 // Other shaders respect forcePPL, but legacy groundcover mods were designed to work with vertex lighting.
@@ -156,11 +158,11 @@ void main(void)
 #endif
 
 #if @diffuseMap
-    diffuseMapUV = (gl_TextureMatrix[@diffuseMapUV] * gl_MultiTexCoord@diffuseMapUV).xy;
+    diffuseMapUV = (texMat@diffuseMapUV * vec4(gl_MultiTexCoord@diffuseMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @normalMap
-    normalMapUV = (gl_TextureMatrix[@normalMapUV] * gl_MultiTexCoord@normalMapUV).xy;
+    normalMapUV = (texMat@normalMapUV * vec4(gl_MultiTexCoord@normalMapUV.xyz, 1.0)).xy;
 #endif
 
 #if PER_PIXEL_LIGHTING
diff --git a/files/shaders/compatibility/gui.vert b/files/shaders/compatibility/gui.vert
index b378b097bd78ac93327bcec7e0c40b6396491359..eefaeb5bd7c754f975c293a63d3630cff0d06dca 100644
--- a/files/shaders/compatibility/gui.vert
+++ b/files/shaders/compatibility/gui.vert
@@ -6,6 +6,6 @@ varying vec4 passColor;
 void main()
 {
     gl_Position = vec4(gl_Vertex.xyz, 1.0);
-    diffuseMapUV = (gl_TextureMatrix[0] * gl_MultiTexCoord0).xy;
+    diffuseMapUV = gl_MultiTexCoord0.xy;
     passColor = gl_Color;
 }
diff --git a/files/shaders/compatibility/objects.vert b/files/shaders/compatibility/objects.vert
index bb8c608b568f79378a745c5f0a7a0fb81b653d29..6b7769bdbc38809f3d33bdbae39e243de3c792ed 100644
--- a/files/shaders/compatibility/objects.vert
+++ b/files/shaders/compatibility/objects.vert
@@ -6,42 +6,52 @@
 
 #if @diffuseMap
 varying vec2 diffuseMapUV;
+uniform mat4 texMat@diffuseMapUV;
 #endif
 
 #if @darkMap
 varying vec2 darkMapUV;
+uniform mat4 texMat@darkMapUV;
 #endif
 
 #if @detailMap
 varying vec2 detailMapUV;
+uniform mat4 texMat@detailMapUV;
 #endif
 
 #if @decalMap
 varying vec2 decalMapUV;
+uniform mat4 texMat@decalMapUV;
 #endif
 
 #if @emissiveMap
 varying vec2 emissiveMapUV;
+uniform mat4 texMat@emissiveMapUV;
 #endif
 
 #if @normalMap
 varying vec2 normalMapUV;
+uniform mat4 texMat@normalMapUV;
 #endif
 
 #if @envMap
 varying vec2 envMapUV;
+uniform mat4 texMat@envMapUV;
 #endif
 
 #if @bumpMap
 varying vec2 bumpMapUV;
+uniform mat4 texMat@bumpMapUV;
 #endif
 
 #if @specularMap
 varying vec2 specularMapUV;
+uniform mat4 texMat@specularMapUV;
 #endif
 
 #if @glossMap
 varying vec2 glossMapUV;
+uniform mat4 texMat@glossMapUV;
 #endif
 
 #define PER_PIXEL_LIGHTING (@normalMap || @specularMap || @forcePPL)
@@ -110,39 +120,39 @@ void main(void)
 #endif
 
 #if @diffuseMap
-    diffuseMapUV = (gl_TextureMatrix[@diffuseMapUV] * gl_MultiTexCoord@diffuseMapUV).xy;
+    diffuseMapUV = (texMat@diffuseMapUV * vec4(gl_MultiTexCoord@diffuseMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @darkMap
-    darkMapUV = (gl_TextureMatrix[@darkMapUV] * gl_MultiTexCoord@darkMapUV).xy;
+    darkMapUV = (texMat@darkMapUV * vec4(gl_MultiTexCoord@darkMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @detailMap
-    detailMapUV = (gl_TextureMatrix[@detailMapUV] * gl_MultiTexCoord@detailMapUV).xy;
+    detailMapUV = (texMat@detailMapUV * vec4(gl_MultiTexCoord@detailMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @decalMap
-    decalMapUV = (gl_TextureMatrix[@decalMapUV] * gl_MultiTexCoord@decalMapUV).xy;
+    decalMapUV = (texMat@decalMapUV * vec4(gl_MultiTexCoord@decalMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @emissiveMap
-    emissiveMapUV = (gl_TextureMatrix[@emissiveMapUV] * gl_MultiTexCoord@emissiveMapUV).xy;
+    emissiveMapUV = (texMat@emissiveMapUV * vec4(gl_MultiTexCoord@emissiveMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @normalMap
-    normalMapUV = (gl_TextureMatrix[@normalMapUV] * gl_MultiTexCoord@normalMapUV).xy;
+    normalMapUV = (texMat@normalMapUV * vec4(gl_MultiTexCoord@normalMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @bumpMap
-    bumpMapUV = (gl_TextureMatrix[@bumpMapUV] * gl_MultiTexCoord@bumpMapUV).xy;
+    bumpMapUV = (texMat@bumpMapUV * vec4(gl_MultiTexCoord@bumpMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @specularMap
-    specularMapUV = (gl_TextureMatrix[@specularMapUV] * gl_MultiTexCoord@specularMapUV).xy;
+    specularMapUV = (texMat@specularMapUV * vec4(gl_MultiTexCoord@specularMapUV.xyz, 1.0)).xy;
 #endif
 
 #if @glossMap
-    glossMapUV = (gl_TextureMatrix[@glossMapUV] * gl_MultiTexCoord@glossMapUV).xy;
+    glossMapUV = (texMat@glossMapUV * vec4(gl_MultiTexCoord@glossMapUV.xyz, 1.0)).xy;
 #endif
 
 #if !PER_PIXEL_LIGHTING
diff --git a/files/shaders/compatibility/shadowcasting.vert b/files/shaders/compatibility/shadowcasting.vert
index c68bdd5c1775a58de3969547a0dcfb9394a31bf1..6b33c5643e56be75469eb6746c00a3dd4b62bd66 100644
--- a/files/shaders/compatibility/shadowcasting.vert
+++ b/files/shaders/compatibility/shadowcasting.vert
@@ -8,6 +8,7 @@ uniform int colorMode;
 uniform bool useTreeAnim;
 uniform bool useDiffuseMapForShadowAlpha = true;
 uniform bool alphaTestShadows = true;
+uniform mat4 texMat0;
 
 void main(void)
 {
@@ -17,7 +18,7 @@ void main(void)
     gl_ClipVertex = viewPos;
 
     if (useDiffuseMapForShadowAlpha)
-        diffuseMapUV = (gl_TextureMatrix[0] * gl_MultiTexCoord0).xy;
+        diffuseMapUV = (texMat0 * vec4(gl_MultiTexCoord0.xyz, 1.0)).xy;
     else
         diffuseMapUV = vec2(0.0); // Avoid undefined behaviour if running on hardware predating the concept of dynamically uniform expressions
     if (colorMode == 2)
diff --git a/files/shaders/compatibility/sky.vert b/files/shaders/compatibility/sky.vert
index 0cd6c0660c32b1f99c7632e90b064d494796a4f8..5f15423e0a0f06af87df04106b26533e0f9a67c5 100644
--- a/files/shaders/compatibility/sky.vert
+++ b/files/shaders/compatibility/sky.vert
@@ -5,6 +5,7 @@
 #include "lib/sky/passes.glsl"
 
 uniform int pass;
+uniform mat4 texMat0;
 
 varying vec4 passColor;
 varying vec2 diffuseMapUV;
@@ -15,7 +16,7 @@ void main()
     passColor = gl_Color;
 
     if (pass == PASS_CLOUDS)
-        diffuseMapUV = (gl_TextureMatrix[0] * gl_MultiTexCoord0).xy;
+        diffuseMapUV = (texMat0 * vec4(gl_MultiTexCoord0.xyz, 1.0)).xy;
     else
         diffuseMapUV = gl_MultiTexCoord0.xy;
 }
diff --git a/files/shaders/compatibility/terrain.frag b/files/shaders/compatibility/terrain.frag
index c4d7d2e8d543427007275eecf145278e035b1001..d691c377707bda2dacd1f111f82a6f238efd0e91 100644
--- a/files/shaders/compatibility/terrain.frag
+++ b/files/shaders/compatibility/terrain.frag
@@ -35,6 +35,9 @@ varying vec3 passNormal;
 uniform vec2 screenRes;
 uniform float far;
 
+uniform mat4 texMat0;
+uniform mat4 texMat1;
+
 #include "lib/core/fragment.h.glsl"
 
 #include "vertexcolors.glsl"
@@ -46,7 +49,7 @@ uniform float far;
 
 void main()
 {
-    vec2 adjustedUV = (gl_TextureMatrix[0] * vec4(uv, 0.0, 1.0)).xy;
+    vec2 adjustedUV = (texMat0 * vec4(uv, 0.0, 1.0)).xy;
 
 #if @parallax
     float height = texture2D(normalMap, adjustedUV).a;
@@ -59,7 +62,7 @@ void main()
     gl_FragData[0].a *= diffuseColor.a;
 
 #if @blendMap
-    vec2 blendMapUV = (gl_TextureMatrix[1] * vec4(uv, 0.0, 1.0)).xy;
+    vec2 blendMapUV = (texMat1 * vec4(uv, 0.0, 1.0)).xy;
     gl_FragData[0].a *= texture2D(blendMap, blendMapUV).a;
 #endif
 
diff --git a/files/shaders/compatibility/terrain_composite.frag b/files/shaders/compatibility/terrain_composite.frag
index ff0b58c8967ad9aa08840b6ed45aabed4abf9923..5dd5e6721033643ddef30c5f2160ed87785ba106 100644
--- a/files/shaders/compatibility/terrain_composite.frag
+++ b/files/shaders/compatibility/terrain_composite.frag
@@ -4,19 +4,22 @@ varying vec2 uv;
 
 uniform sampler2D diffuseMap;
 
+uniform mat4 texMat0;
+uniform mat4 texMat1;
+
 #if @blendMap
 uniform sampler2D blendMap;
 #endif
 
 void main()
 {
-    vec2 adjustedUV = (gl_TextureMatrix[0] * vec4(uv, 0.0, 1.0)).xy;
+    vec2 adjustedUV = (texMat0 * vec4(uv, 0.0, 1.0)).xy;
 
     vec4 diffuseTex = texture2D(diffuseMap, adjustedUV);
     gl_FragData[0] = vec4(diffuseTex.xyz, 1.0);
 
 #if @blendMap
-    vec2 blendMapUV = (gl_TextureMatrix[1] * vec4(uv, 0.0, 1.0)).xy;
+    vec2 blendMapUV = (texMat1 * vec4(uv, 0.0, 1.0)).xy;
     gl_FragData[0].a *= texture2D(blendMap, blendMapUV).a;
 #endif
 }
