diff --git a/components/sceneutil/shadow.cpp b/components/sceneutil/shadow.cpp
index 0e6abe3258..29d17b9b75 100644
--- a/components/sceneutil/shadow.cpp
+++ b/components/sceneutil/shadow.cpp
@@ -87,6 +87,9 @@ namespace SceneUtil
         if (!mEnableShadows)
             return;
 
+        stateset.setDefine("DISABLE_SHADOWS", "1", osg::StateAttribute::ON | osg::StateAttribute::OVERRIDE | osg::StateAttribute::PROTECTED);
+            return;
+
         osg::ref_ptr<osg::Image> fakeShadowMapImage = new osg::Image();
         fakeShadowMapImage->allocateImage(1, 1, 1, GL_DEPTH_COMPONENT, GL_FLOAT);
         *(float*)fakeShadowMapImage->data() = std::numeric_limits<float>::infinity();
