diff --git a/apps/openmw/mwrender/postprocessor.cpp b/apps/openmw/mwrender/postprocessor.cpp
index ff9ac2a42e..dcdc1b59e4 100644
--- a/apps/openmw/mwrender/postprocessor.cpp
+++ b/apps/openmw/mwrender/postprocessor.cpp
@@ -195,17 +195,15 @@ namespace MWRender
         if (!ext->glDisablei && ext->glDisableIndexedEXT)
             ext->glDisablei = ext->glDisableIndexedEXT;
 
-#ifdef ANDROID
-        ext->glDisablei = nullptr;
-#endif
-
         if (ext->glDisablei)
             mNormalsSupported = true;
         else
             Log(Debug::Error) << "'glDisablei' unsupported, pass normals will not be available to shaders.";
 
-        mGLSLVersion = static_cast<int>(ext->glslLanguageVersion * 100);
-        mUBO = ext->isUniformBufferObjectSupported && mGLSLVersion >= 330;
+//        mGLSLVersion = static_cast<int>(ext->glslLanguageVersion * 100);
+//        mUBO = ext->isUniformBufferObjectSupported && mGLSLVersion >= 330;
+        mGLSLVersion = (getenv("OPENMW_FORCEGLSL330") == nullptr) ? 120 : 330;
+        mUBO = (getenv("OPENMW_UBO") == nullptr) ? false : true;
         mStateUpdater = new Fx::StateUpdater(mUBO);
 
         addChild(mHUDCamera);
@@ -412,11 +410,14 @@ namespace MWRender
 
             if (mNormalsSupported)
             {
+/*
                 auto& shaderManager
                     = MWBase::Environment::get().getResourceSystem()->getSceneManager()->getShaderManager();
                 auto defines = shaderManager.getGlobalDefines();
                 defines["disableNormals"] = mNormals ? "0" : "1";
                 shaderManager.setGlobalDefines(defines);
+*/
+                mRendering.setWriteNormals(mNormals);
             }
 
             mRendering.getLightRoot()->setCollectPPLights(mPassLights);
diff --git a/apps/openmw/mwrender/renderingmanager.cpp b/apps/openmw/mwrender/renderingmanager.cpp
index 21c25cd7b1..4625aa51fd 100644
--- a/apps/openmw/mwrender/renderingmanager.cpp
+++ b/apps/openmw/mwrender/renderingmanager.cpp
@@ -1303,17 +1303,19 @@ namespace MWRender
             else if (it->first == "Shaders"
                 && (it->second == "force per pixel lighting" || it->second == "classic falloff"))
             {
+/*
                 mViewer->stopThreading();
 
                 auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
                 defines["forcePPL"] = Settings::shaders().mForcePerPixelLighting ? "1" : "0";
                 defines["classicFalloff"] = Settings::shaders().mClassicFalloff ? "1" : "0";
                 mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                 if (MWMechanics::getPlayer().isInCell() && it->second == "classic falloff")
                     configureAmbient(*MWMechanics::getPlayer().getCell()->getCell());
 
-                mViewer->startThreading();
+//                mViewer->startThreading();
+                mStateUpdater->reset();
             }
             else if (it->first == "Shaders"
                 && (it->second == "light bounds multiplier" || it->second == "maximum light distance"
@@ -1326,18 +1328,19 @@ namespace MWRender
 
                 if (it->second == "max lights")
                 {
-                    mViewer->stopThreading();
 
-                    lightManager->updateMaxLights(Settings::shaders().mMaxLights);
+//                    mViewer->stopThreading();
 
+                    lightManager->updateMaxLights(Settings::shaders().mMaxLights);
+/*
                     auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
                     for (const auto& [name, key] : lightManager->getLightDefines())
                         defines[name] = key;
                     mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                     mStateUpdater->reset();
 
-                    mViewer->startThreading();
+//                    mViewer->startThreading();
                 }
             }
             else if (it->first == "Post Processing" && it->second == "enabled")
@@ -1586,4 +1589,9 @@ namespace MWRender
     {
         mNavMesh->setMode(value);
     }
+
+    void RenderingManager::setWriteNormals(bool enabled)
+    {
+        mStateUpdater->setWriteNormals(enabled);
+    }
 }
diff --git a/apps/openmw/mwrender/renderingmanager.hpp b/apps/openmw/mwrender/renderingmanager.hpp
index b8ad535eb3..a6e04698b6 100644
--- a/apps/openmw/mwrender/renderingmanager.hpp
+++ b/apps/openmw/mwrender/renderingmanager.hpp
@@ -279,6 +279,8 @@ namespace MWRender
 
         void setNavMeshMode(Settings::NavMeshRenderMode value);
 
+        void setWriteNormals(bool enabled);
+
     private:
         void updateTextureFiltering();
         void updateAmbient();
diff --git a/components/shader/shadervisitor.cpp b/components/shader/shadervisitor.cpp
index a9ad457a67..2c70536861 100644
--- a/components/shader/shadervisitor.cpp
+++ b/components/shader/shadervisitor.cpp
@@ -677,7 +677,10 @@ namespace Shader
             defineMap["endLight"] = "0";
 
         if (simpleLighting || dynamic_cast<osgParticle::ParticleSystem*>(&node))
-            defineMap["forcePPL"] = "0";
+        {
+             defineMap["forcePPL"] = "0";
+            writableStateSet->setDefine("FORCE_PPL", "0", osg::StateAttribute::ON | osg::StateAttribute::PROTECTED);
+        }
 
         bool particleOcclusion = false;
         node.getUserValue("particleOcclusion", particleOcclusion);
